cue/cuex/runtime/packagemanager.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/kubevela/pkg/util/k8s"
+	"github.com/kubevela/pkg/util/maps"
+	"github.com/kubevela/pkg/util/singleton"
+	"github.com/kubevela/pkg/util/slices"
+)
+
+const defaultResyncPeriod = 5 * time.Minute
+
+// PackageManager manages cue packages
+type PackageManager struct {
+	Internals *maps.SyncMap[string, Package]
+	Externals *maps.SyncMap[string, Package]
+
+	ResyncPeriod time.Duration
+	StopCh       chan struct{}
+}
+
+type PackageManagerOption interface {
+	ApplyTo(*PackageManager)
+}
+
+type WithResyncPeriod time.Duration
+
+func (in WithResyncPeriod) ApplyTo(m *PackageManager) {
+	m.ResyncPeriod = time.Duration(in)
+}
+
+type WithInternalPackage struct {
+	Package
+}
+
+func (in WithInternalPackage) ApplyTo(m *PackageManager) {
+	m.Internals.Set(in.GetPath(), in)
+}
+
+// NewPackageManager create PackageManager with given options
+func NewPackageManager(opts ...PackageManagerOption) *PackageManager {
+	m := &PackageManager{
+		Internals:    maps.NewSyncMap[string, Package](),
+		Externals:    maps.NewSyncMap[string, Package](),
+		ResyncPeriod: defaultResyncPeriod,
+	}
+	for _, opt := range opts {
+		opt.ApplyTo(m)
+	}
+	return m
+}
+
+func (in *PackageManager) getExternalPackageID(pkg *v1alpha1.Package) string {
+	return "external://" + pkg.GetNamespace() + "/" + pkg.GetName()
+}
+
+func (in *PackageManager) setExternalPackage(pkg *v1alpha1.Package) {
+	_id := in.getExternalPackageID(pkg)
+	_pkg, err := NewExternalPackage(pkg)
+	if err != nil {
+		klog.Errorf("parse external package %s/%s failed: %s", pkg.Namespace, pkg.Name, err.Error())
+		return
+	}
+	in.Externals.Set(_id, _pkg)
+}
+
+func (in *PackageManager) delExternalPackage(pkg *v1alpha1.Package) {
+	_id := in.getExternalPackageID(pkg)
+	in.Externals.Del(_id)
+}
+
+// LoadExternalPackages load all external packages
+func (in *PackageManager) LoadExternalPackages(ctx context.Context) error {
+	pkgs, err := singleton.DynamicClient.Get().Resource(v1alpha1.PackageGroupVersionResource).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, pkg := range pkgs.Items {
+		_pkg := &v1alpha1.Package{}
+		if err = apiruntime.DefaultUnstructuredConverter.FromUnstructured(pkg.Object, _pkg); err != nil {
+			return err
+		}
+		in.setExternalPackage(_pkg)
+	}
+	return nil
+}
+
+// ListenExternalPackages start informer to listen external package changes
+func (in *PackageManager) ListenExternalPackages(stopCh <-chan struct{}) {
+	if stopCh == nil && in.StopCh == nil {
+		in.StopCh = make(chan struct{})
+		stopCh = in.StopCh
+	}
+	factory := dynamicinformer.NewDynamicSharedInformerFactory(
+		singleton.DynamicClient.Get(), in.ResyncPeriod)
+	informer := factory.ForResource(v1alpha1.PackageGroupVersionResource).Informer()
+	defer runtime.HandleCrash()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if o, err := k8s.AsStructured[v1alpha1.Package](obj.(*unstructured.Unstructured)); err == nil {
+				in.setExternalPackage(o)
```

Comment:
What if the external package has the same path with the internal package?

Reply:
external package has higher priority than internal package currently